pisa/stages/test_bourdeet/add_indices.py

```diff
@@ -0,0 +1,136 @@
+    ----------
+    data
+    params
+        foo : Quantity
+        bar : Quanitiy with time dimension
+    input_names
+    output_names
+    debug_mode
+    input_specs: 
+    calc_specs : must be events
+    output_specs: must be a MultiDimBinnig
+
+    Notes:
+    ------
+
+    - input and calc specs are predetermined in the module
+        (inputs from the config files will be disregarded)
+
+    - stage appends an array quantity called bin_indices
+    - stage also appends an array mask to access events by
+      bin index later in the pipeline
+
+    """
+
+    # this is the constructor with default arguments
+    
+    def __init__(self,
+                 data=None,
+                 params=None,
+                 input_names=None,
+                 output_names=None,
+                 debug_mode=None,
+                 input_specs=None,
+                 calc_specs=None,
+                 output_specs=None,
+                ):
+
+        #
+        # No parameters are expected in this stage
+        # same goes for a bunch of other stage options
+        #
+        expected_params = ()
+        input_names = ()
+        output_names = ()
+        input_apply_keys = ()
+
+        # We add the bin_indices key
+        # (but not in the apply function so maybe useless...)
+        #
+        output_calc_keys = ('bin_indices',)
+        output_apply_keys = ()
+
+        # init base class
+        super(add_indices, self).__init__(data=data,
+                                       params=params,
+                                       expected_params=expected_params,
+                                       input_names=input_names,
+                                       output_names=output_names,
+                                       debug_mode=debug_mode,
+                                       input_specs=input_specs,
+                                       calc_specs=calc_specs,
+                                       output_specs=output_specs,
+                                       input_apply_keys=input_apply_keys,
+                                       output_apply_keys=output_apply_keys,
+                                       output_calc_keys=output_calc_keys,
+                                       )
+
+        # make sure the user specified some modes
+        assert self.input_mode is not None
+        assert self.calc_mode is not None
+        assert self.output_mode is not None
+
+    def setup_function(self):
+        '''
+        Calculate the bin index where each event falls into
+
+        Create one mask for each analysis bin.
+        '''
+        
+        assert self.calc_specs == 'events', 'ERROR: calc specs must be set to "events for this module'
+
+        self.data.data_specs = 'events'
+
+        for container in self.data:
+            # Generate a new container called bin_indices
+            container['bin_indices'] = np.empty((container.size), dtype=FTYPE)
+  
+            e = container['reco_energy']
```

Comment:
you should make this adapt to the binning; this is hard-coded to work only with specific binning spec (even changing the order will break this code). (I believe output specs should be able to tell you what you need to do this dynamically.)

Reply:
I've modified that line to be more generic now